narwhal/storage/src/proposer_store.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use store::rocks::open_cf;
+use store::{reopen, rocks::DBMap, Map};
+use types::{Header, StoreResult};
+
+pub type ProposerKey = u32;
+
+pub const LAST_PROPOSAL_KEY: ProposerKey = 0;
+
+/// The storage for the proposer
+#[derive(Clone)]
+pub struct ProposerStore {
+    /// Holds the Last Header that was proposed by the Proposer.
+    last_proposed: DBMap<ProposerKey, Header>,
+}
+
+impl ProposerStore {
+    pub fn new(last_proposed: DBMap<ProposerKey, Header>) -> ProposerStore {
+        Self { last_proposed }
```

Comment:
I think this would be the natural place indeed to store own signed headers beyond the very last one. Then upon a commit we look at all the headers in lower rounds (<- check), and if the headers are not included in the commit, we re-include their batch digests in future headers?

Reply:
can't we already do that with the certificate store? (although we would need to parse all the certificates that have been created from others)

Reply:
Yes, that was the idea @gdanezis